apycula/chipdb.py

```diff
@@ -1575,6 +1577,60 @@ def fse_bram(fse, aux = False):
     bels[name] = Bel()
     return bels
 
+def fse_dsp(fse, aux = False):
+    bels = {}
+    if aux:
+        bels['DSP_AUX0'] = Bel()
+        bels['DSP_AUX1'] = Bel()
+    else:
+        # These are two macro DSPs, their purpose is to manage the control
+        # signals CE, CLK and RESET, which seem to be allocated to different
+        # subblocks from a common pool, the size of which reaches 4 possible
+        # PIPs for each type of signal.
+        # In other words, only the portmap that describes the pool is important here.
+        bels['DSP'] = Bel()
+        bels['DSP0'] = Bel()  # Macro 0
+        bels['DSP1'] = Bel()  # Macro 1
+        # Padd
+        bels['PADD900'] = Bel()  # macro 0 padd9 0
+        bels['PADD901'] = Bel()  # macro 0 padd9 1
+        bels['PADD902'] = Bel()  # macro 0 padd9 2
+        bels['PADD903'] = Bel()  # macro 0 padd9 3
+        bels['PADD1800'] = Bel() # macro 0 padd18 0
+        bels['PADD1801'] = Bel() # macro 0 padd18 1
+        bels['PADD910'] = Bel()  # macro 1 padd9 0
+        bels['PADD911'] = Bel()  # macro 1 padd9 1
+        bels['PADD912'] = Bel()  # macro 1 padd9 2
+        bels['PADD913'] = Bel()  # macro 1 padd9 3
+        bels['PADD1810'] = Bel() # macro 1 padd18 0
+        bels['PADD1811'] = Bel() # macro 1 padd18 1
+        # mult
+        bels['MULT9X900'] = Bel()   # macro 0 mult9x9 0
+        bels['MULT9X901'] = Bel()   # macro 0 mult9x9 1
+        bels['MULT9X902'] = Bel()   # macro 0 mult9x9 2
+        bels['MULT9X903'] = Bel()   # macro 0 mult9x9 3
+        bels['MULT18X1800'] = Bel() # macro 0 mult18x18 0
+        bels['MULT18X1801'] = Bel() # macro 0 mult18x18 1
+        bels['MULT9X910'] = Bel()   # macro 1 mult9x9 0
+        bels['MULT9X911'] = Bel()   # macro 1 mult9x9 1
+        bels['MULT9X912'] = Bel()   # macro 1 mult9x9 2
+        bels['MULT9X913'] = Bel()   # macro 1 mult9x9 3
+        bels['MULT18X1810'] = Bel() # macro 1 mult18x18 0
+        bels['MULT18X1811'] = Bel() # macro 1 mult18x18 1
+        # alu
+        bels['ALU54D0'] = Bel()     # macro 0 ALU54D
+        bels['ALU54D1'] = Bel()     # macro 1 ALU54D
+        # multalu
+        bels['MULTALU18X180'] = Bel()     # macro 0 multalu 18x18
+        bels['MULTALU18X181'] = Bel()     # macro 1 multalu 18x18
+        bels['MULTALU36X180'] = Bel()     # macro 0 multalu 36x18
+        bels['MULTALU36X181'] = Bel()     # macro 1 multalu 36x18
+        bels['MULTADDALU18X180'] = Bel()     # macro 0 multaddalu 18x18
+        bels['MULTADDALU18X181'] = Bel()     # macro 1 multaddalu 18x18
+
+        bels['MULT36X36'] = Bel()   # entire DSP mult36x36
+
+    return bels
```

Comment:
Do the DSP blocks just not have any modes or flags at all? Just fixed function blocks with inputs and outputs? Hm interesting!

Reply:
For the most part this is true - separate blocks with their own inputs and outputs. We glue them together like a 4-d multiplier to get Mult36x36.

![dsp](https://github.com/YosysHQ/apicula/assets/6075465/d0dedd0d-7bbb-4760-aa63-46ed842b4c43)